File: src/Scanner.cpp
#include <iostream>
#include <cctype>
#include "Scanner.h"

namespace bigNumCalculator {

Scanner::Scanner(const std::string& content)
   : _source(content.begin(), content.end()),
     _source_size{(int)content.size()}
{
    if (content.size() == 0) {
        return;
    }

    _current = _source[0];
    _offset = 0;
}

Word Scanner::scan() {
    skip_whitespace();

    switch (_current) {
        case EOFCHAR:
            return Word{Token::END, std::vector<char> { }};
        case '+':
            next();
            return Word{Token::ADD, std::vector<char> { '+' }};
        case '*':
            next();
            return Word{Token::MUL, std::vector<char> { '*' }};
        case '-':
            next();
            return Word{Token::SUB, std::vector<char> { '-' }};
        case '/':
            next();
            return Word{Token::DIV, std::vector<char> { '/' }};
        case '%':
            next();
            return Word{Token::MOD, std::vector<char> { '%' }};
        case '(':
            next();
            return Word{Token::LPARENT, std::vector<char> { '(' }};
        case ')':
            next();
            return Word{Token::RPARENT, std::vector<char> { ')' }};
        case '^':
            next();
            return Word{Token::POWER, std::vector<char> { '^' }};
    }

    if (std::isdigit(_current)) {
        return scan_number();
    }

    return Word{Token::ILLEGAL, std::vector<char> { _current }};
}

// scan_number scans numbers - integers and real numbers. Real number can be
// expressed in decimal or scientific notation.
Word Scanner::scan_number() {
    auto start_position = _offset;
    auto prev_char = _current;
    auto token = Token::INT;

    for (;;) {
        next();

        if (!std::isdigit(_current) && _current != '.' && _current != 'e'
                && _current != 'E' && _current != '+' && _current != '-') {
            break;
        }

        if (_current == '.') {
            token = Token::REAL;
        }

        if ((_current == '+' || _current == '-') && (prev_char != 'e' || prev_char != 'E')) {
            break;
        }

        prev_char = _current;
    }

    auto start = _source.begin() + start_position;
    auto end = _source.begin() + _offset;
    std::vector<char> number(start, end);

    return Word{token, number};
}

void Scanner::next() {
    if (_offset + 1 >= (int)_source.size()) {
        _offset = _source_size;
        _current = EOFCHAR;
        return;
    }

    _offset++;
    _current = _source[_offset];
}

void Scanner::skip_whitespace() {
    while (std::isspace(_source[_offset])) {
        next();
    }
}

} // bigNumCalculator namespace

File: src/bnCalc/Scanner.cs
using System;
using System.Collections.Generic;
using bnCalc.Models;

namespace bnCalc
{
    public interface IScanner
    {
        List<Word> Scan();
    }

    public class Scanner : IScanner
    {
        private readonly string _source;
        private char _current;
        private int _offset = -1;

        public Scanner(string content)
        {
            if (content is null || content.Length == 0)
            {
                _source = "";
                return;
            }

            _source = content;
            _current = _source[0];
            _offset = 0;
        }

        public List<Word> Scan()
        {
            var words = new List<Word>();

            if (_offset == -1)
            {
                return words;
            }

            for (;;)
            {
                var word = ScanSingle();

                if (word.Token == Token.Illegal)
                {
                    throw new ArgumentException($"Got illegal sequence of characters starting from '{word.Chars?.ToString()}'.");
                }

                if (word.Token == Token.End)
                {
                    return words;
                }

                words.Add(word);
            }
        }

        private Word ScanSingle()
        {
            SkipWhiteSpace();

            switch (_current)
            {
                case char.MinValue:
                    return new(Token.End, null);
                case '+':
                    Next();
                    return new(Token.Add, new List<char> { '+' });
                case '-':
                    Next();
                    return new(Token.Sub, new List<char> { '-' });
                case '*':
                    Next();
                    return new(Token.Mul, new List<char> { '*' });
                case '/':
                    Next();
                    return new(Token.Div, new List<char> { '/' });
                case '%':
                    Next();
                    return new(Token.Mod, new List<char> { '%' });
                case '^':
                    Next();
                    return new(Token.Power, new List<char> { '^' });
                case '(':
                    Next();
                    return new(Token.LParent, new List<char> { '(' });
                case ')':
                    Next();
                    return new(Token.RParent, new List<char> { ')' });
            }

            if (char.IsDigit(_current))
            {
                return ScanNumber();
            }

            return new(Token.Illegal, new List<char> { _current });
        }

        private Word ScanNumber()
        {
            var chars = new List<char>();
            var prevChar = _current;
            var token = Token.Int;

            for (;;)
            {
                if (!char.IsDigit(_current) && _current != '.' &&
                    char.ToLower(_current) != 'e' && _current != '+' && _current != '-')
                {
                    break;
                }

                if (_current == '.')
                {
                    token = Token.Real;
                }

                if ((_current == '-' || _current == '+') && (char.ToLower(prevChar) != 'e')) {
                    break;
                }

                prevChar = _current;
                chars.Add(_current);
                Next();
            }

            return new(token, chars);
        }

        private void SkipWhiteSpace()
        {
            while (char.IsWhiteSpace(_current))
            {
                Next();
            }
        }

        private void Next()
        {
            if (_offset + 1 >= _source.Length)
            {
                _offset = _source.Length;
                _current = char.MinValue;
                return;
            }

            _offset++;
            _current = _source[_offset];
        }
    }
}
